### PR TITLE
AEA-1362 ML Skill Bug fix

### PR DESCRIPTION
**Pull request by 5A11**
This PR cleans up and fixes the logic of ML Trade skills.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/fetchai/skills/ml_train/handlers.py [hunk 1]
```diff
@@ -201,7 +201,7 @@ def _handle_data(self, ml_trade_msg: MlTradeMessage) -> None:
             )
             self.context.task_manager.enqueue_task(
                 MLTrainTask(
-                    skill_context=self.context, train_data=data[:2], epochs_per_batch=5,
+                    skill_context=self.context, train_data=data[:2],
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/handlers.py:204]: This should only receive the weights and the data and any configuration needed. No context!

File: packages/fetchai/skills/ml_train/tasks.py [hunk 2]
```diff
@@ -35,46 +37,34 @@ def __init__(
         skill_context: SkillContext,
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/tasks.py:37]: drop skill context

File: packages/fetchai/skills/ml_train/tasks.py [hunk 3]
```diff
@@ -35,46 +37,34 @@ def __init__(
...
         super().__init__(logger=skill_context.logger)
-        self.train_x, self.train_y = train_data
 
-        self.model = self._make_model()
+        self.train_x, self.train_y = train_data
+        self.weights = weights
         self.epochs_per_batch = epochs_per_batch
-        self.batch_size = batch_size
+
+        self.ml_model = cast(MLModel, skill_context.ml_model)
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/tasks.py:49]: cannot do that because we don't have context anymore

File: packages/fetchai/skills/ml_train/tasks.py [hunk 4]
```diff
@@ -35,46 +37,34 @@ def __init__(
...
 
     def setup(self) -> None:
         """Set up the task."""
         self.logger.info("ML Train task: setup method called.")
 
-    @staticmethod
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/tasks.py:52]: we keep that

File: packages/fetchai/skills/ml_train/tasks.py [hunk 5]
```diff
@@ -35,46 +37,34 @@ def __init__(
...
-        )
-        return model
-
     def execute(self, *args: Any, **kwargs: Any) -> Any:
         """Execute the task."""
         self.logger.info("Start training with {} rows".format(self.train_x.shape[0]))
-        self.model.fit(self.train_x, self.train_y, epochs=self.epochs_per_batch)
-        loss, acc = self.model.evaluate(self.train_x, self.train_y, verbose=2)
+
+        model = self.ml_model.make_model(self.weights)
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/tasks.py:59]: 1. make model from static method above
2. set weight on model from weight added in constructor
3. we call fit
4. get weights
5. print loss, acc
6. we return the new weights

Comment (DavidMinarsch) [reply packages/fetchai/skills/ml_train/tasks.py:59]: so only 1 needs to change

File: packages/fetchai/skills/ml_train/handlers.py [hunk 6]
```diff
@@ -201,7 +201,7 @@ def _handle_data(self, ml_trade_msg: MlTradeMessage) -> None:
             )
             self.context.task_manager.enqueue_task(
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/handlers.py:202]: get your task_id, save it locally, then use it in behaviour to check if task is done

File: packages/fetchai/skills/ml_train/behaviours.py [hunk 7]
```diff
@@ -42,7 +43,26 @@
...
+        """
+        strategy = cast(Strategy, self.context.strategy)
+
+        if strategy.current_task_id is not None:
+            result = self.context.task_manager.get_task_result(strategy.current_task_id)
+
+            if not result.successful():
+                return
+
+            strategy.weights = result.get()
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/behaviours.py:63]: task id should be set to None after getting result successfully.

File: packages/fetchai/skills/ml_train/strategy.py [hunk 8]
```diff
@@ -121,6 +124,26 @@ def balance(self, balance: int) -> None:
...
+        """Get the current_task_id."""
+        return self._current_task_id
+
+    @current_task_id.setter
+    def current_task_id(self, task_id: int) -> None:
+        """Set the current_task_id."""
+        self._current_task_id = task_id
+
+    @property
+    def weights(self) -> Optional[Any]:
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/strategy.py:138]: Try to find out what the type is of the weights.

File: tests/test_packages/test_skills/test_ml_train/test_task.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+"""This module contains the tests of the task class of the ml_train skill."""
+
+import sys
+from pathlib import Path
+from typing import Tuple
+from unittest.mock import patch
+
+import numpy as np
+import pytest
+import tensorflow as tf
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_ml_train/test_task.py:28]: Lazy import to avoid test failures

File: packages/fetchai/skills/ml_train/behaviours.py [hunk 10]
```diff
@@ -42,7 +43,31 @@
...
+            if not result.ready():
+                return
+
+            if not result.successful():
+                return
+
+            ml_task = result.get()
+            strategy.weights = ml_task.result
+            strategy.current_task_id = None
+
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/behaviours.py:69]: Here we should add:
```
if len(strategy.data) > 0:
      data = strateg.data.pop(0)
      ml_task_id = self.context.task_manager.enqueue_task(
                 MLTrainTask(
                     train_data=data[:2], epochs_per_batch=5, weights=strategy.weights
                 )
             )
             strategy.current_task_id = ml_task_id
            return
```

Comment (DavidMinarsch) [reply packages/fetchai/skills/ml_train/behaviours.py:69]: To ensure, we only ever train one task at a time.

File: packages/fetchai/skills/ml_train/handlers.py [hunk 11]
```diff
@@ -199,11 +200,12 @@ def _handle_data(self, ml_trade_msg: MlTradeMessage) -> None:
...
                 )
             )
-            self.context.task_manager.enqueue_task(
+            ml_task_id = self.context.task_manager.enqueue_task(
                 MLTrainTask(
-                    skill_context=self.context, train_data=data[:2], epochs_per_batch=5,
+                    train_data=data[:2], epochs_per_batch=5, weights=strategy.weights
                 )
             )
+            self.context.strategy.current_task_id = ml_task_id
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/handlers.py:208]: Change to `strategy.data.add(data)`

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Review by jrriehl — APPROVED** [review]
LGTM
